test/test_stream.py

```diff
@@ -0,0 +1,112 @@
+def handler_from(directory):
+    def _init(self, *args, **kwargs):
+        return SimpleHTTPRequestHandler.__init__(
+            self, *args, directory=self.directory, **kwargs
+        )  # type: ignore[call-arg]
+
+    return type(f"HandlerFrom<{directory}>", (SimpleHTTPRequestHandler,), {"__init__": _init, "directory": directory},)
+
+
+class TestStream(expecttest.TestCase):
+    def setUp(self) -> None:
+        # Temporary Directory abd Server Setup
+        print("\nSetup is running...")
+        self.temp_dir = tempfile.TemporaryDirectory()  # noqa: P201
+        self.temp_dir_path = self.temp_dir.name
+        self.port = 8006
+        self.stop_server = False
+        self.server_thread = threading.Thread(
+            target=self.running_server
+        )  # TestStream.start_test_server(self.temp_dir_path, self.port)
+        self.server_thread.start()
+
+    def tearDown(self) -> None:
+        print("Tear down is running...")
+        self.stop_server = True
+        self.server_thread.join(timeout=3)
+        print("Server is stopped")
+        print("Removing Temp Directory...")
+        self.temp_dir.cleanup()
+        print("Tear down has completed")
+
+    def running_server(self):
+        httpd = socketserver.TCPServer(("", self.port), handler_from(self.temp_dir_path))
+        print(f"Starting TCP server on port {self.port}...")
+        httpd.serve_forever()
+        while True:
+            if self.stop_server:  # TODO: This is not closing
+                httpd.server_close()
+
+    @staticmethod
+    def generate_binary_files(path, num_files: int, num_kbs: int = 16) -> List[str]:
+        print("Generating binary files...")
+        ls = []
+        for i in range(num_files):
+            with tempfile.NamedTemporaryFile(dir=path, delete=False, prefix=f"{i}", suffix=".byte") as f:
+                temp_binary_file_name = os.path.join(path, f.name)
+            with open(temp_binary_file_name, "wb") as binary_file:
+                binary_file.write(b"0123456789abcdef" * 64 * num_kbs)
+            ls.append(temp_binary_file_name)
+        return ls
+
+    @staticmethod
+    def get_name_mode_from_file_type(file_type: str) -> Tuple[str, str, str]:
+        if file_type == "tar":  # Uncompressed
+            return "test_tar.tar", "w:tar", "r|tar"  # Must use '|' instead of ':' to get a stream
+        elif file_type == "gz":
+            return "test_tar.tar.gz", "w:gz", "r|gz"
+        elif file_type == "xz":
+            return "test_tar.tar.xz", "w:xz", "r|xz"
+        else:
+            raise ValueError(f"file_type must be 'tar', 'gz', or 'xz', but instead found {file_type}.")
+
+    @staticmethod
+    def add_to_tar_archive(file_path, write_mode: str, file_names: List[str]) -> None:
+        print("Adding files to Tar archive...")
+        with tarfile.open(file_path, write_mode) as tar:
+            for fname in file_names:
+                tar.add(fname)
```

Comment:
You may need to specify `arcname` for `tar.add` to override the file name, because `tar.add` would preserve the hierarchy of each file. Let's say your archive is `/tmp/foo/1.tar`, and file is `/tmp/foo/1.csv`. If you don't specify the `arcname`, the file is going to be `/tmp/foo/1.tar/tmp/foo/1.cdv` when it's extracted from tar.